src/index.d.ts

```diff
@@ -1,966 +1,54 @@
-export function createSelector<S1, S2, S3, S4, S5, S6, S7, S8, S9, S10, S11, S12, P1, P2, P3, P4, P5, P6, P7, P8, P9, P10, P11, P12, R1, R2, R3, R4, R5, R6, R7, R8, R9, R10, R11, R12, T>(
-  selectors: [ParametricSelector<S1, P1, R1>, ParametricSelector<S2, P2, R2>, ParametricSelector<S3, P3, R3>, ParametricSelector<S4, P4, R4>, ParametricSelector<S5, P5, R5>, ParametricSelector<S6, P6, R6>, ParametricSelector<S7, P7, R7>, ParametricSelector<S8, P8, R8>, ParametricSelector<S9, P9, R9>, ParametricSelector<S10, P10, R10>, ParametricSelector<S11, P11, R11>, ParametricSelector<S12, P12, R12>],
-  combiner: (res1: R1, res2: R2, res3: R3, res4: R4, res5: R5, res6: R6, res7: R7, res8: R8, res9: R9, res10: R10, res11: R11, res12: R12) => T,
-): OutputParametricSelector<S1 & S2 & S3 & S4 & S5 & S6 & S7 & S8 & S9 & S10 & S11 & S12, P1 & P2 & P3 & P4 & P5 & P6 & P7 & P8 & P9 & P10 & P11 & P12, T, (res1: R1, res2: R2, res3: R3, res4: R4, res5: R5, res6: R6, res7: R7, res8: R8, res9: R9, res10: R10, res11: R11, res12: R12) => T>;
-
-/* any number of uniform selectors */
-export function createSelector<S, R, T>(
-  selectors: Selector<S, R>[],
-  combiner: (...res: R[]) => T,
-): OutputSelector<S, T, (...res: R[]) => T>;
-export function createSelector<S, P, R, T>(
-  selectors: ParametricSelector<S, P, R>[],
-  combiner: (...res: R[]) => T,
-): OutputParametricSelector<S, P, T, (...res: R[]) => T>;
+};
+
+type SelectorArray = ReadonlyArray<Selector>
+
+type GetStateFromSelector<S> = S extends Selector<infer State> ? State : never;
+type GetStateFromSelectors<S extends SelectorArray> =
+  S extends [infer Current, ...infer Other]
+    ? Current extends Selector
+      ? Other extends SelectorArray
+        ? GetStateFromSelector<Current> | GetStateFromSelectors<Other>
+        : GetStateFromSelector<Current>
+      : never
+    : S extends (infer Elem)[] ? GetStateFromSelector<Elem> : never;
+
+type GetParamsFromSelector<S> = S extends Selector<any, any, infer P> ? P extends [] ? never : P : never;
+export type GetParamsFromSelectors<S, Found = never> = S extends SelectorArray
+  ? S extends (infer s)[]
+    ? GetParamsFromSelector<s>
+    : S extends [infer Current, ...infer Rest]
+      ? GetParamsFromSelector<Current> extends []
+        ? GetParamsFromSelectors<Rest, Found>
+        : GetParamsFromSelector<Current>
+      : S
+  : Found;
+
+type SelectorResultArray<Selectors extends SelectorArray, Rest extends SelectorArray = Selectors> =
+  Rest extends [infer S, ...infer Remaining]
+    ? S extends Selector
+      ? Remaining extends SelectorArray
+        ? [ReturnType<S>, ...SelectorResultArray<Selectors, Remaining>]
+        : [ReturnType<S>]
+    : []
+  : Rest extends ((...args: any) => infer S)[] ? S[] : [];
+
+export function createSelector<Selectors extends SelectorArray, Result>(
```

Comment:
The breaking change in the arguments to `createSelector` would be kinda expected for me - I'd be okay with that.